### PR TITLE
[#171184989] Add client certificate validation

### PR DESCRIPTION
**Pull request by francescopersico**

### CODE REVIEW

File: src/App.ts [hunk 1]
```diff
@@ -17,6 +17,7 @@ import { specs as publicApiV1Specs } from "../generated/api/public_api_pagopa";
 import { Configuration } from "./Configuration";
 import { GetOpenapi } from "./controllers/openapi";
 import * as PaymentController from "./controllers/restful/PaymentController";
+import { requireClientCertificateFingerprint } from "./midlewares/requireClientCertificateFingerprint";
```
Comment (gunzip) [inline src/App.ts]:
```suggestion
import { requireClientCertificateFingerprint } from "./middlewares/requireClientCertificateFingerprint";
```

File: src/midlewares/requireClientCertificateFingerprint.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+      if (
+        clientCertificateBase64 !== undefined &&
+        clientCertificateBase64 !== ""
+      ) {
+        const pem = `-----BEGIN CERTIFICATE-----${clientCertificateBase64}-----END CERTIFICATE-----`;
+        const clientCertificate = pki.certificateFromPem(pem);
+
+        // Extract the fingerprint
+        const fingerprint = md.sha256
```
Comment (gunzip) [inline src/midlewares/requireClientCertificateFingerprint.ts]: if this is an expensive operation you may consider memoizing it for the same value of
`clientCertificateBase64`

File: src/midlewares/requireClientCertificateFingerprint.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+          .update(
+            asn1.toDer(pki.certificateToAsn1(clientCertificate)).getBytes()
+          )
+          .digest()
+          .toHex()
+          .toUpperCase();
+
+        // Very the fingerprint
+        if (fingerprint !== validFingerprint) {
+          res.status(403).send("Invalid client certificate");
```
Comment (gunzip) [inline src/midlewares/requireClientCertificateFingerprint.ts]: return ?

File: src/midlewares/types.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import * as express from "express";
```
Comment (gunzip) [inline src/midlewares/types.ts]: have you checked if this type exists into @types/express and is something like express.Hander ?

**Comment by francescopersico** [comment]
> is pagopa-proxy exposed on internet ? because in this case, since this library does not implement cache eviction, I guess is it possible to exhaust memory sending multiple certs as input (maybe I'm wrong but needs to be checked)

io-pagopa-proxy will be hosted on App Service with a rule that allow only connection from pagoPA node.

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>Please include a description of your PR changes.</td>
    </tr>
  </tbody>
</table>



## Affected stories

  * 🌟 [#171184989](https://www.pivotaltracker.com/story/show/171184989): Il pagopa-proxy deve supportare la verifica del cliente certificate


New dependencies added: <a href='https://www.npmjs.com/package/fast-memoize'><code>fast-memoize</code></a>, <a href='https://www.npmjs.com/package/node-forge'><code>node-forge</code></a> and <a href='https://www.npmjs.com/package/@types/node-forge'><code>&#64;types/node-forge</code></a>.


<h2><a href="https://github.com/caiogondim/fast-memoize#readme">fast-memoize</a></h2>
<p>Author: Caio Gondim</p>
<p>Description: Fastest memoization lib that supports N arguments</p>
<p>Homepage: <a href="https://github.com/caiogondim/fast-memoize#readme">https://github.com/caiogondim/fast-memoize#readme</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>almost 4 years ago</td></tr><tr><td>Last Updated</td><td>over 1 year ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>22</td></tr><tr><td>Direct Dependencies</td><td></td></tr>
</table>

<details>
<summary><code>README</code></summary></br>



# fast-memoize

<div>
  Travis CI  Code coverage <a href="https://www.npmjs.com/package/fast-memoize"></a>
</div>

<br>

> In computing, memoization is an optimization technique used primarily to speed up computer programs by storing the results of expensive function calls and returning the cached result when the same inputs occur again.
> — Wikipedia

This library is an attempt to make the **fastest possible memoization library in
JavaScript that supports *N* arguments**.

## Installation

```shell
npm install fast-memoize --save
```

## Usage

```js
const memoize = require('fast-memoize')

const fn = function (one, two, three) { /* ... */ }

const memoized = memoize(fn)

memoized('foo', 3, 'bar')
memoized('foo', 3, 'bar') // Cache hit
```

### Custom cache

The fastest cache is used for the running environment, but it is possible to
pass a custom cache to be used.

```js
const memoized = memoize(fn, {
  cache: {
    create() {
      var store = {};
      return {
        has(key) { return (key in store); },
        get(key) { return store[key]; },
        set(key, value) { store[key] = value; }
      };
    }
  }
})
```

The custom cache should be an object containing a `create` method that returns
an object implementing the following methods:
- `get`
- `set`
- `has`

### Custom serializer

To use a custom serializer:
```js
const memoized = memoize(fn, {
  serializer: customSerializer
})
```

The serializer is a function that receives one argument and outputs a string
that represents it. It has to be a
[deterministic algorithm](https://en.wikipedia.org/wiki/Deterministic_algorithm)
meaning that, given one input, it always returns the same output.

## Benchmark

For an in depth explanation on how this library was created, go read
[this post on RisingStack](https://community.risingstack.com/the-worlds-fastest-javascript-memoization-library/).

Below you can see a performance benchmark between some of the most popular libraries
for memoization.



To run the benchmark, clone the repo, install the dependencies and run `npm run benchmark`.
```shell
git clone git@github.com:caiogondim/fast-memoize.git
cd fast-memoize
npm install
npm run benchmark
```



### Against another git hash

To benchmark the current code against a git hash, branch, ...

```shell
npm run benchmark:compare 53fa9a62214e816cf8b5b4fa291c38f1d63677b9
```

### Gotchas

#### Spread arguments

We check for `function.length` to get upfront the expected number of arguments
in order to use the fastest strategy. But with spread arguments we don't receive
the right number.

```js
function multiply (multiplier, ...theArgs) {
  return theArgs.map(function (element) {
    return multiplier * element
  })
}
multiply.length // => 1
```

So if you use spread arguments, explicitly set the strategy to variadic.

```js
const memoizedMultiply = memoize(multiply, {
  strategy: memoize.strategies.variadic
})
```

#### Function Arguments

The default serializer uses `JSON.stringify` which will serialize functions as
`null`. This means that if you are passing any functions as arguments you will
get the same output regardless of whether you pass in different functions or
indeed no function at all. The cache key generated will always be the same. To
get around this you can give each function a unique ID and use that.

```js
let id = 0
function memoizedId(x) {
  if (!x.__memoizedId) x.__memoizedId = ++id
  return { __memoizedId: x.__memoizedId }
}

memoize((aFunction, foo) => {
  return aFunction.bind(foo)
}, {
  serializer: args => {
    const argumentsWithFuncIds = Array.from(args).map(x => {
      if (typeof x === 'function') return memoizedId(x)
      return x
    })
    return JSON.stringify(argumentsWithFuncIds)
  }
})
```

## Credits
- Icon by Mary Rankin from the Noun Project
- [Bullet train ZSH theme](https://github.com/caiogondim/bullet-train-oh-my-zsh-theme)

---

[caiogondim.com](https://caiogondim.com) &nbsp;&middot;&nbsp;
GitHub [@caiogondim](https://github.com/caiogondim) &nbsp;&middot;&nbsp;
Twitter [@caio_gondim](https://twitter.com/caio_gondim)


</details>




<h2><a href="https://github.com/digitalbazaar/forge">node-forge</a></h2>
<p>Author: Digital Bazaar, Inc.</p>
<p>Description: JavaScript implementations of network transports, cryptography, ciphers, PKI, message digests, and various utilities.</p>
<p>Homepage: <a href="https://github.com/digitalbazaar/forge">https://github.com/digitalbazaar/forge</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 6 years ago</td></tr><tr><td>Last Updated</td><td>5 months ago</td></tr><tr><td>License</td><td>(BSD-3-Clause OR GPL-2.0)</td></tr><tr><td>Maintainers</td><td>3</td></tr><tr><td>Releases</td><td>123</td></tr><tr><td>Keywords</td><td>aes, asn, asn.1, cbc, crypto, cryptography, csr, des, gcm, hmac, http, https, md5, network, pkcs, pki, prng, rc2, rsa, sha1, sha256, sha384, sha512, ssh, tls, x.509 and x509</td></tr>
</table>
This README is too long to show.



<h2><a href="http://npmjs.com/package/@types/node-forge">&#64;types/node-forge</a></h2>
<p>Author: Unknown</p>
<p>Description: TypeScript definitions for node-forge</p>
<p>Homepage: <a href="http://npmjs.com/package/@types/node-forge">http://npmjs.com/package/@types/node-forge</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 3 years ago</td></tr><tr><td>Last Updated</td><td>6 days ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>35</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/@types/node'><code>&#64;types/node</code></a></td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# Installation
> `npm install --save @types/node-forge`

# Summary
This package contains type definitions for node-forge (https://github.com/digitalbazaar/forge).

# Details
Files were exported from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/node-forge.

### Additional Details
 * Last updated: Tue, 04 Feb 2020 21:45:08 GMT
 * Dependencies: [@types/node](https://npmjs.com/package/@types/node)
 * Global values: none

# Credits
These definitions were written by Seth Westphal    (https://github.com/westy92), Kay Schecker     (https://github.com/flynetworks), Aakash Goenka    (https://github.com/a-k-g), Rafal2228        (https://github.com/rafal2228), Beeno Tung       (https://github.com/beenotung), Joe Flateau      (https://github.com/joeflateau), Nikita Koryabkin (https://github.com/Apologiz), timhwang21       (https://github.com/timhwang21), supaiku0         (https://github.com/supaiku0), Anders Kaseorg   (https://github.com/andersk), Sascha Zarhuber  (https://github.com/saschazar21), Rogier Schouten  (https://github.com/rogierschouten), and Ivan Aseev       (https://github.com/aseevia).


</details>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against ab5d56cb91804b62253e2c948c27d6f55d4c77b1
</p>

**Pull request merged**